internal/testutils/jimmtest/e2e_suite.go

```diff
@@ -142,9 +147,65 @@ func (s *WebsocketE2ESuite) SetUpTest(c *gc.C) {
 	s.APIHandler = mux
 	s.HTTP = httptest.NewTLSServer(s.APIHandler)
 
+	s.AddAdminUser(c, "alice@canonical.com")
+
+	cct := names.NewCloudCredentialTag(TestE2ECloudName + "/charlie@canonical.com/cred")
+	cloudCredentials := s.GetExistingClientCredentialsForCloud(c, TestE2ECloudName)
+	s.UpdateCloudCredential(c, cct, cloudCredentials)
+	s.Credential2 = new(dbmodel.CloudCredential)
+	s.Credential2.SetTag(cct)
+	err := s.JIMM.Database.GetCloudCredential(ctx, s.Credential2)
+	c.Assert(err, gc.Equals, nil)
+	mt := s.AddModel(c, names.NewUserTag("charlie@canonical.com"), petname.Generate(2, "-"), names.NewCloudTag(TestE2ECloudName), TestE2ECloudRegionName, cct)
+	s.Model2 = new(dbmodel.Model)
+	s.Model2.SetTag(mt)
+	err = s.JIMM.Database.GetModel(ctx, s.Model2)
+	c.Assert(err, gc.Equals, nil)
+	mt = s.AddModel(c, names.NewUserTag("charlie@canonical.com"), petname.Generate(2, "-"), names.NewCloudTag(TestE2ECloudName), TestE2ECloudRegionName, cct)
+	s.Model3 = new(dbmodel.Model)
+	s.Model3.SetTag(mt)
+	err = s.JIMM.Database.GetModel(ctx, s.Model3)
+	c.Assert(err, gc.Equals, nil)
+
+	bobIdentity, err := dbmodel.NewIdentity("bob@canonical.com")
+	c.Assert(err, gc.IsNil)
+
+	bob := openfga.NewUser(
+		bobIdentity,
+		s.OFGAClient,
+	)
+	err = bob.SetModelAccess(ctx, s.Model3.ResourceTag(), ofganames.ReaderRelation)
+	c.Assert(err, gc.Equals, nil)
+}
+
+func (s *WebsocketE2ESuite) DeployApplication(c *gc.C, user *openfga.User, modelTag names.Tag, appName string) {
+	modelTagConv, ok := modelTag.(names.ModelTag)
+	c.Assert(ok, gc.Equals, true)
+
+	conn := s.Open(c, nil, user.Name, &modelTagConv)
+	defer conn.Close()
+
+	client := application.NewClient(conn)
+
+	_, _, errs := client.DeployFromRepository(application.DeployFromRepositoryArg{
+		CharmName:       "juju-qa-test",
+		ApplicationName: appName,
+		Channel:         utils.Ptr("latest/stable"),
+		NumUnits:        utils.Ptr(1),
+	})
+	for err := range errs {
+		c.Assert(err, gc.Equals, nil)
+	}
+
 }
 
 func (s *WebsocketE2ESuite) TearDownTest(c *gc.C) {
+	if s.Model2 != nil {
+		s.DestroyModelAndDeleteFromDatabase(c, s.Model2.ResourceTag())
+	}
+	if s.Model3 != nil {
+		s.DestroyModelAndDeleteFromDatabase(c, s.Model3.ResourceTag())
+	}
```

Comment:
Avoid these and just run cleanup based on what was set during setup and test creation.